src/uct/ib/mlx5/ib_mlx5.c

```diff
@@ -67,23 +67,55 @@ ucs_config_field_t uct_ib_mlx5_iface_config_table[] = {
      "+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++\n",
      ucs_offsetof(uct_ib_mlx5_iface_config_t, ar_enable), UCS_CONFIG_TYPE_TERNARY_AUTO},
 
-    {"DISABLE_CQE_ZIPPING", "y",
-     "Disable the possibility of CQE zipping\n",
-     ucs_offsetof(uct_ib_mlx5_iface_config_t, disable_cqe_zipping), UCS_CONFIG_TYPE_BOOL},
+    {"CQE_ZIPPING_ENABLE", "no",
+     "Enable CQE zipping feature.\n"
+     "The CQE zipping reduces the PCI load by merging and compressing the almost identical CQEs to compression blocks.\n"
+     "In some cases this feature can have negative performance impact.",
+     ucs_offsetof(uct_ib_mlx5_iface_config_t, cqe_zipping_enable), UCS_CONFIG_TYPE_TERNARY},
 
     {NULL}
 };
 
+#if HAVE_DECL_MLX5DV_CQ_INIT_ATTR_MASK_CQE_SIZE
+static UCS_F_ALWAYS_INLINE ucs_status_t
+uct_ib_mlx5_handle_cqe_zipping_mode(int cqe_zipping_enable,
+                                    struct mlx5dv_cq_init_attr* dv_attr,
+                                    uint32_t md_flags)
+{
+    if (cqe_zipping_enable == UCS_NO) {
+        return UCS_OK;
+    }
+
+    if (((dv_attr->cqe_size == 64)  && (md_flags & UCT_IB_MLX5_MD_FLAG_CQE64_ZIP)) ||
+        ((dv_attr->cqe_size == 128) && (md_flags & UCT_IB_MLX5_MD_FLAG_CQE128_ZIP))) {
+        dv_attr->comp_mask |= MLX5DV_CQ_INIT_ATTR_MASK_COMPRESSED_CQE;
+        dv_attr->cqe_comp_res_format = MLX5DV_CQE_RES_FORMAT_CSUM;
+        return UCS_OK;
+    }
+
+    if (cqe_zipping_enable == UCS_YES) {
+        ucs_error("The CQE_ZIPPING_ENABLE option set to \"yes\", but this "
+                  "feature is not supported by current HW configuration.");
+        return UCS_ERR_UNSUPPORTED;
+    }
+
+    return UCS_OK;
+}
+#endif
+
 ucs_status_t uct_ib_mlx5_create_cq(uct_ib_iface_t *iface, uct_ib_dir_t dir,
+                                   const uct_ib_mlx5_iface_config_t* mlx5_config,
+                                   const uct_ib_iface_config_t* ib_config,
```

Comment:
What do you propose?

`mlx5_config` is required to check the `cqe_zipping_enable` 

`ib_config` is used to pass it to the `uct_ib_verbs_create_cq()`. This argument doesn't used inside this function, but it exists there for consistency (this function should be consistent with the `create_cq` function pointer).

And as I know we cannot get the `ib_config` from the `iface`. Also we cannot cast `ib_config` to `mlx5_config` and vise versa since they have no inheritance relationship and have different structures dependent on type of connection (ud/rc/dc).

Reply:
discussed offline